### DISCUSSION TITLE
How to use RavenDB time series to store missing values of temperature, along with other measure for the same timestamp?

### DISCUSSION THREAD

**Discussion by pmengal** [ravendb/ravendb#19582] [Q&A] [answered]
In the context of multiple metrics for a given timestamp, how do you deal with the fact that RavenDB cannot store null values for some missing metrics? For example, the value 0 for the temperature measurement does not mean that there is no temperature measurement. We also have telemetry with different recording frequencies.

So how do you store a non-zero value to indicate the absence of a measurement (missing values) without resorting to workarounds such as using an outlier (e.g. -99999)? What techniques have you used successfully in your project?

Thanks in advance.

**Comment by karmeli87** [comment]
You are correct, currently there is no option to append a "missing" value to the time-series. 

The reason behind it, is that we consider multi-dimensional time-series to represent a single entry. For example, coordinates where there is no meaning for one of the axis being missing. (Also the internal implementation rely on that assumption, for example what would be the `Length` of the  time-series that has some values only in some dimensions missing?)

You can have multiple time-series per document. Would it make sense for you maybe to have a time-series per telemetry?

**Reply by ayende** [reply]
Note that you can add _less_ elements.
So let say that you have `X, Y` where `Y` is optional.
You can add an entry with both elements, and then an entry with just one.

Beyond that, you cannot add `NaN` or similar to have a "hole" in the middle.

Otherwise, you can work with multiple time series for this.

**Reply by pmengal** [reply]
Thanks for your responses. Using one time serie per metric also overcome the limitation of 32 mesures which was a problem for us. Performance wise, what would be the impact in case we go that way? If we have 32 metrics, we'll have to do 32 queries right?

**Reply by ayende** [reply]
No, you can include all of the time series in one queri

**Reply by karmeli87** [reply, marked as answer]
Regarding the queries, I guess it depends on what you have right now. You can index the time-series with map, multi-map or [map-all](https://ravendb.net/docs/article-page/6.2/csharp/document-extensions/timeseries/indexing#map-index---index-all-time-series-from-single-collection).  
Another option is to query them directly.

here is an example of how to query 2 time-series
```
public class Machine
{
    public string Id;
    public bool Active;
    public string Name;
}

private struct Temperature
{
    [TimeSeriesValue(0)] public double Celsius;
}

[RavenTheory(RavenTestCategory.TimeSeries | RavenTestCategory.Querying)]
[RavenData(DatabaseMode = RavenDatabaseMode.Single)]
public void QuerySeveralTimeSeries(Options options)
{
    using (var store = GetDocumentStore(options))
    {
        using (var session = store.OpenSession())
        {
            session.Store(new Machine
            {
                Name = "The Incredible Machine",
                Active = true
            }, "m/1");

            for (int i = 0; i < 100; i++)
            {
                session.TimeSeriesFor<Temperature>("m/1","temp1").Append(DateTime.UtcNow.AddSeconds(i * 3), new Temperature
                {
                    Celsius = 30 + i
                });

                session.TimeSeriesFor<Temperature>("m/1","temp2").Append(DateTime.UtcNow.AddSeconds(i * 5), new Temperature
                {
                    Celsius = -14 - i
                });
            }

            session.SaveChanges();
        }

        using (var session = store.OpenSession())
        {
            var query = session.Query<Machine>()
                .Where(m => m.Active)
                .Select(m => new
                {
                    T1 = RavenQuery.TimeSeries<Temperature>(m, "temp1")
                        .GroupBy(t => t.Minutes(1))
                        .Select(g => new
                        {
                            Avg = g.Average(),
                            Max = g.Max()
                        })
                        .ToList(),
                    T2 = RavenQuery.TimeSeries<Temperature>(m, "temp2")
                        .GroupBy(t => t.Minutes(1))
                        .Select(g => new
                        {
                            Avg = g.Average(),
                            Max = g.Max()
                        })
                        .ToList(),
                });

            var result = query.ToList();

        }
    }
}
```

**Comment by pmengal** [comment]
Thanks a million
